[Site Isolation] Connect scrolling trees from main process and iframe process

nmoucht opened the pull request:
#### e7e9773f128b368bf877a9f1481917a7350e9b05
<pre>
[Site Isolation] Connect scrolling trees from main process and iframe process
<a href="https://bugs.webkit.org/show_bug.cgi?id=268591">https://bugs.webkit.org/show_bug.cgi?id=268591</a>
<a href="https://rdar.apple.com/122145583">rdar://122145583</a>

Reviewed by Simon Fraser.

This is part 2 of splitting up <a href="https://github.com/WebKit/WebKit/pull/23242">https://github.com/WebKit/WebKit/pull/23242</a>
This connects the scrolling node marked with the hosting id with the root
of a transaction marked with the same id. Since the scrolling state tree
only includes state changes, we need to keep track of all state trees committed
with a host id if we haven&apos;t received a commit with the corresponding host id
in the tree.

* LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree.html: Added.
* LayoutTests/http/tests/site-isolation/scrolling/resources/empty-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/scrolling/resources/overflow-scroll-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/scrolling/scrolling-tree-iframe-overflow-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/scrolling-tree-iframe-overflow.html: Added.
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::createNode):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::mainFrameIdentifier const):
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::ScrollingStateTree):
(WebCore::ScrollingStateTree::insertNode):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
(WebCore::ScrollingStateTree::frameIdentifier const):
(WebCore::ScrollingStateTree::setFrameIdentifier):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::removeNode):
(WebCore::ScrollingTree::commitTreeState):
(WebCore::ScrollingTree::updateTreeFromStateNodeRecursive):
(WebCore::ScrollingTree::removeAllNodes):
(WebCore::ScrollingTree::addScrollingNodeToHostedSubtreeMap):
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::isScrollingSynchronizedWithMainThread):
(WebCore::ScrollingTree::treeLock):
(WebCore::ScrollingTree::propagateSynchronousScrollingReasons):
(WebCore::ScrollingTree::WTF_REQUIRES_LOCK): Deleted.
(WebCore::ScrollingTree::WTF_RETURNS_LOCK): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.cpp:
(WebCore::ScrollingTreeFrameHostingNode::setLayerHostingContextIdentifier):
(WebCore::ScrollingTreeFrameHostingNode::removeHostedChildren):
(WebCore::ScrollingTreeFrameHostingNode::willBeDestroyed):
(WebCore::ScrollingTreeFrameHostingNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.h:
(WebCore::ScrollingTreeFrameHostingNode::addHostedChild):
(WebCore::ScrollingTreeFrameHostingNode::setLayerHostingContextIdentifier): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeNode.cpp:
(WebCore::ScrollingTreeNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingTreeNode.h:
(WebCore::ScrollingTreeNode::isRootOfHostedSubtree const):
(WebCore::ScrollingTreeNode::frameIdentifier const):
(WebCore::ScrollingTreeNode::setFrameIdentifier):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::dumpProperties const):
* Source/WebCore/page/scrolling/ThreadedScrollingTree.cpp:
(WebCore::ThreadedScrollingTree::waitForEventToBeProcessedByMainThread):
(WebCore::ThreadedScrollingTree::waitForRenderingUpdateCompletionOrTimeout):
* Source/WebCore/page/scrolling/ThreadedScrollingTree.h:
(WebCore::ThreadedScrollingTree::treeLock):
(WebCore::ThreadedScrollingTree::WTF_RETURNS_LOCK): Deleted.
(WebCore::ThreadedScrollingTree::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeFrameScrollingNodeMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(WebKit::RemoteScrollingCoordinatorTransaction::RemoteScrollingCoordinatorTransaction):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.h:
(WebKit::RemoteScrollingCoordinatorTransaction::RemoteScrollingCoordinatorTransaction):
(WebKit::RemoteScrollingCoordinatorTransaction::frameIdentifier const):
(WebKit::RemoteScrollingCoordinatorTransaction::setFrameIdentifier):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::commitScrollingTreeState):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
(WebKit::RemoteScrollingTree::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::waitForEventDefaultHandlingCompletion):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::buildTransaction):

Canonical link: <a href="https://commits.webkit.org/274918@main">https://commits.webkit.org/274918@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff892d1a2753aa3cac2883d2d90743f1ae10b65d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40366 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42911 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36455 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42673 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16707 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33485 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16325 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34804 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14066 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14123 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35768 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44189 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36587 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36091 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39831 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15181 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12430 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38105 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16800 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/35051 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16849 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5347 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16444 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -801,6 +801,9 @@ void AsyncScrollingCoordinator::scrollableAreaScrollbarLayerDidChange(Scrollable
 ScrollingNodeID AsyncScrollingCoordinator::createNode(ScrollingNodeType nodeType, ScrollingNodeID newNodeID)
 {
     LOG_WITH_STREAM(ScrollingTree, stream << "AsyncScrollingCoordinator::createNode " << nodeType << " node " << newNodeID);
+    // TODO: Need a better way to fix scrolling tree in iframe process
```
Ideally file a bug to track the fix and reference it here.

nmoucht replied:
Added.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
```diff
@@ -60,6 +60,7 @@ static void nodeWasReattachedRecursive(ScrollingStateNode& node)
 
 ScrollingStateTree::ScrollingStateTree(AsyncScrollingCoordinator* scrollingCoordinator)
     : m_scrollingCoordinator(scrollingCoordinator)
+, m_frameIdentifier(FrameIdentifier { })
```
Bad indentation. Can we not get this identifier at ScrollingStateTree construction time? Leaving it unset opens you up to bugs.

nmoucht replied:
Fixed. Unfortunately we can't get at the FrameIdentifier from here. We set it after every transaction so I don't think we have to worry about it.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
```diff
@@ -220,12 +221,11 @@ ScrollingNodeID ScrollingStateTree::insertNode(ScrollingNodeType nodeType, Scrol
 
     RefPtr<ScrollingStateNode> newNode;
     if (!parentID) {
-        RELEASE_ASSERT(nodeType == ScrollingNodeType::MainFrame);
```
Can we assert MainFrame || Subframe?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateTree.h:
```diff
@@ -111,6 +113,8 @@ class ScrollingStateTree : public CanMakeCheckedPtr {
     void traverse(const ScrollingStateNode&, const Function<void(const ScrollingStateNode&)>&) const;
 
     ThreadSafeWeakPtr<AsyncScrollingCoordinator> m_scrollingCoordinator;
+    FrameIdentifier m_frameIdentifier;
```
Maybe m_rootFrameIdentifier or whatever we call the root-most frame in an isolated process.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -56,6 +57,14 @@ struct CommitTreeState {
+    // Hosted subtrees needing attaching to scrolling tree after main commit has finished
+    Vector<std::pair<LayerHostingContextIdentifier, Vector<std::unique_ptr<ScrollingStateTree>>>> pendingSubtreesNeedingCommit;
+    // Nodes that are descendants of a frame hosting node.
+    HashSet<ScrollingNodeID> hostedScrollingNodes;
```
The name should communicate that these are all subframe nodes.

nmoucht replied:
So these are just the list of all descendants of the frameHostingNode not just subframe nodes.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -56,6 +57,14 @@ struct CommitTreeState {
+    // Nodes that are descendants of a frame hosting node.
+    HashSet<ScrollingNodeID> hostedScrollingNodes;
+    // This has a value when doing a commit for a hosted subtree
+    RefPtr<ScrollingTreeFrameHostingNode> hostingNode;
```
frameHostingNode

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -56,6 +57,14 @@ struct CommitTreeState {
+    // This has a value when doing a commit for a hosted subtree
+    RefPtr<ScrollingTreeFrameHostingNode> hostingNode;
+    // Identifier for the parent frame associated with this commit
+    FrameIdentifier frameId;
```
"parent frame" gives me pause. Is it the parent frame, or this frame? How does it related to hostingNode?

nmoucht replied:
Right this is just the frame associated with the commit not the parent.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -309,17 +318,40 @@ void ScrollingTree::setOverlayScrollbarsEnabled(bool enabled)
+    }
+}
 
+bool ScrollingTree::commitTreeStateInternal(std::unique_ptr<ScrollingStateTree>&& scrollingStateTree, std::optional<LayerHostingContextIdentifier> identifier)
```
`identifier` -> `hostingContextIdentifier`

smfr replied:
Also, whose identifier is this? Does it go along with the scrollingStateTree which might be from a subframe?

nmoucht replied:
Fixed. This has a value if it comes from the subframe process.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -309,17 +318,40 @@ void ScrollingTree::setOverlayScrollbarsEnabled(bool enabled)
+{
     bool rootStateNodeChanged = scrollingStateTree->hasNewRootStateNode();
 
     LOG(ScrollingTree, "\nScrollingTree %p commitTreeState", this);
```
I think much more logging about subframe tree connections would be useful.

nmoucht replied:
Right will re-add the logging I had locally.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -309,17 +318,40 @@ void ScrollingTree::setOverlayScrollbarsEnabled(bool enabled)
+            commitState.hostingNode = scrollingNode;
+        else {
+            // Add to hosted subtrees needing pairing since the hosting node is not in the scrolling tree yet
+            Vector<std::unique_ptr<ScrollingStateTree>> vect;
```
`vent` -> `stateTreesPendingConnection` or something.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -309,17 +318,40 @@ void ScrollingTree::setOverlayScrollbarsEnabled(bool enabled)
+            // Add to hosted subtrees needing pairing since the hosting node is not in the scrolling tree yet
+            Vector<std::unique_ptr<ScrollingStateTree>> vect;
+            if (m_hostedSubtreesNeedingPairing.contains(*identifier))
+            vect = m_hostedSubtreesNeedingPairing.take(*identifier);
```
Bad indentation

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -352,14 +384,25 @@ bool ScrollingTree::commitTreeState(std::unique_ptr<ScrollingStateTree>&& scroll
+    auto nodesForFrame = m_nodeMapPerFrame.get(commitState.frameId);
+    for (auto& nodes : nodesForFrame)
+        commitState.unvisitedNodes.add(nodes->scrollingNodeID());
     bool succeeded = updateTreeFromStateNodeRecursive(rootNode.get(), commitState);
```
Blank line here please

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -406,8 +460,12 @@ bool ScrollingTree::updateTreeFromStateNodeRecursive(const ScrollingStateNode* s
+        }
         m_nodeMap.set(nodeID, node.get());
+        m_nodeMapPerFrame.ensure(state.frameId, [] { return HashSet<RefPtr<ScrollingTreeNode>> { }; }).iterator->value.add(node);
+        ALWAYS_LOG_WITH_STREAM(stream << "m_nodeMapPerFrame frameID: " << m_nodeMapPerFrame.size());
```
Remove or turn into normal logging.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -456,6 +517,17 @@ bool ScrollingTree::updateTreeFromStateNodeRecursive(const ScrollingStateNode* s
         state.synchronousScrollingNodes.add(nodeID);
 #endif
+
+    if (is<ScrollingTreeFrameHostingNode>(*node)) {
```
Use `dynamicDowncast<>` here

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -922,6 +995,10 @@ Seconds ScrollingTree::maxAllowableRenderingUpdateDurationForSynchronization()
     return allowableFrameFraction * frameDuration();
 }
 
+void ScrollingTree::addScrollingNodeToHostedSubtreeMap(WebCore::LayerHostingContextIdentifier identifier, RefPtr<ScrollingTreeFrameHostingNode> node)
```
Don't need `WebCore::`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.h:
```diff
@@ -50,6 +55,7 @@ class ScrollingTreeFrameHostingNode : public ScrollingTreeNode {
     WEBCORE_EXPORT void dumpProperties(WTF::TextStream&, OptionSet<ScrollingStateTreeAsTextBehavior>) const override;
 
     std::optional<LayerHostingContextIdentifier> m_hostingContext;
+    HashSet<RefPtr<ScrollingTreeNode>> m_hostedChildren;
```
Can you use a stronger type than `ScrollingTreeNode` ? Are they all frame scrolling nodes?

achristensen07 replied:
Does this need to be able to hold null?  Can we use Ref instead of RefPtr?

nmoucht replied:
No these can be any scrolling tree node. Right this can be Ref.

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeNode.h:
```diff
@@ -109,6 +114,7 @@ class ScrollingTreeNode : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr
 
     const ScrollingNodeType m_nodeType;
     const ScrollingNodeID m_nodeID;
+    FrameIdentifier m_parentFrameIdentifier;
```
Can't this live on ScrollingTreeFrameScrollingNode?

nmoucht replied:
No we need a way to go from ScrollingNode -> FrameID for any node.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.h:
```diff
@@ -60,6 +63,7 @@ class RemoteScrollingCoordinatorTransaction {
     // Data encoded here should be "imperative" (valid just for one transaction). Stateful things should live on scrolling tree nodes.
     // Maybe RequestedScrollData should move here.
     bool m_clearScrollLatching { false };
+    WebCore::FrameIdentifier m_frameID;
```
Which frame? the root frame in the subprocess? A more expressive name would help.

nmoucht replied:
Changed to root frame.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -283,9 +283,7 @@
-        // FIXME: Making scrolling trees work with site isolation.
-        if (layerTreeTransaction.isMainFrameProcessTransaction())
-            requestedScroll = webPageProxy->scrollingCoordinatorProxy()->commitScrollingTreeState(scrollingTreeTransaction);
+        requestedScroll = webPageProxy->scrollingCoordinatorProxy()->commitScrollingTreeState(scrollingTreeTransaction, layerTreeTransaction.remoteContextHostedIdentifier());
```
Is this tested?

nmoucht replied:
No added a radar to test this bit.

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -309,17 +318,40 @@ void ScrollingTree::setOverlayScrollbarsEnabled(bool enabled)
+            // Add to hosted subtrees needing pairing since the hosting node is not in the scrolling tree yet
+            Vector<std::unique_ptr<ScrollingStateTree>> vect;
+            if (m_hostedSubtreesNeedingPairing.contains(*identifier))
+            vect = m_hostedSubtreesNeedingPairing.take(*identifier);
```
This is a double hash lookup.  Use find instead of contains/take.

nmoucht replied:
Hmm I'm not sure there's a great way to do this as the take is necessary since the value is a unique_ptr.

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingTree.h:
```diff
@@ -334,6 +343,9 @@ class ScrollingTree : public ThreadSafeRefCounted<ScrollingTree> {
     MonotonicTime m_lastWheelEventTime WTF_GUARDED_BY_LOCK(m_lastWheelEventTimeLock);
 
+    HashMap<LayerHostingContextIdentifier, Vector<std::unique_ptr<ScrollingStateTree>>> m_hostedSubtreesNeedingPairing;
+    HashMap<LayerHostingContextIdentifier, RefPtr<ScrollingTreeFrameHostingNode>> m_hostedSubtrees;
```
Can the value be a Ref?

nmoucht replied:
Fixed.

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingTree.h:
```diff
@@ -296,6 +303,8 @@ class ScrollingTree : public ThreadSafeRefCounted<ScrollingTree> {
     using ScrollingTreeNodeMap = HashMap<ScrollingNodeID, RefPtr<ScrollingTreeNode>>;
     ScrollingTreeNodeMap m_nodeMap;
 
+    HashMap<FrameIdentifier, HashSet<RefPtr<ScrollingTreeNode>>> m_nodeMapPerFrame;
```
Can the value be a Ref?

nmoucht replied:
Fixed.

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
```diff
@@ -60,6 +60,7 @@ static void nodeWasReattachedRecursive(ScrollingStateNode& node)
 
 ScrollingStateTree::ScrollingStateTree(AsyncScrollingCoordinator* scrollingCoordinator)
     : m_scrollingCoordinator(scrollingCoordinator)
+    , m_rootFrameIdentifier(FrameIdentifier { })
```
This is probably unnecessary.  FrameIdentifier already default initializes.  If anything, we might want to make it std::optional to indicate there can be a state when it hasn't been set yet.

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -56,6 +57,14 @@ struct CommitTreeState {
     // orphanNodes keeps child nodes alive while we rebuild child lists.
     OrphanScrollingNodeMap orphanNodes;
+    // Hosted subtrees needing attaching to scrolling tree after main commit has finished
+    Vector<std::pair<LayerHostingContextIdentifier, Vector<std::unique_ptr<ScrollingStateTree>>>> pendingSubtreesNeedingCommit;
```
Do we need the possibility of having null ScrollingStateTrees, or can this be a UniqueRef<ScrollingStateTree>?

nmoucht committed:
[Site Isolation] Connect scrolling trees from main process and iframe process
https://bugs.webkit.org/show_bug.cgi?id=268591
rdar://122145583

Reviewed by Simon Fraser.

This is part 2 of splitting up https://github.com/WebKit/WebKit/pull/23242
This connects the scrolling node marked with the hosting id with the root
of a transaction marked with the same id. Since the scrolling state tree
only includes state changes, we need to keep track of all state trees committed
with a host id if we haven't received a commit with the corresponding host id
in the tree.

* LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree.html: Added.
* LayoutTests/http/tests/site-isolation/scrolling/resources/empty-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/scrolling/resources/overflow-scroll-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/scrolling/scrolling-tree-iframe-overflow-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/scrolling-tree-iframe-overflow.html: Added.
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::createNode):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::mainFrameIdentifier const):
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::ScrollingStateTree):
(WebCore::ScrollingStateTree::insertNode):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
(WebCore::ScrollingStateTree::frameIdentifier const):
(WebCore::ScrollingStateTree::setFrameIdentifier):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::removeNode):
(WebCore::ScrollingTree::commitTreeState):
(WebCore::ScrollingTree::updateTreeFromStateNodeRecursive):
(WebCore::ScrollingTree::removeAllNodes):
(WebCore::ScrollingTree::addScrollingNodeToHostedSubtreeMap):
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::isScrollingSynchronizedWithMainThread):
(WebCore::ScrollingTree::treeLock):
(WebCore::ScrollingTree::propagateSynchronousScrollingReasons):
(WebCore::ScrollingTree::WTF_REQUIRES_LOCK): Deleted.
(WebCore::ScrollingTree::WTF_RETURNS_LOCK): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.cpp:
(WebCore::ScrollingTreeFrameHostingNode::setLayerHostingContextIdentifier):
(WebCore::ScrollingTreeFrameHostingNode::removeHostedChildren):
(WebCore::ScrollingTreeFrameHostingNode::willBeDestroyed):
(WebCore::ScrollingTreeFrameHostingNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.h:
(WebCore::ScrollingTreeFrameHostingNode::addHostedChild):
(WebCore::ScrollingTreeFrameHostingNode::setLayerHostingContextIdentifier): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeNode.cpp:
(WebCore::ScrollingTreeNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingTreeNode.h:
(WebCore::ScrollingTreeNode::isRootOfHostedSubtree const):
(WebCore::ScrollingTreeNode::frameIdentifier const):
(WebCore::ScrollingTreeNode::setFrameIdentifier):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::dumpProperties const):
* Source/WebCore/page/scrolling/ThreadedScrollingTree.cpp:
(WebCore::ThreadedScrollingTree::waitForEventToBeProcessedByMainThread):
(WebCore::ThreadedScrollingTree::waitForRenderingUpdateCompletionOrTimeout):
* Source/WebCore/page/scrolling/ThreadedScrollingTree.h:
(WebCore::ThreadedScrollingTree::treeLock):
(WebCore::ThreadedScrollingTree::WTF_RETURNS_LOCK): Deleted.
(WebCore::ThreadedScrollingTree::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeFrameScrollingNodeMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(WebKit::RemoteScrollingCoordinatorTransaction::RemoteScrollingCoordinatorTransaction):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.h:
(WebKit::RemoteScrollingCoordinatorTransaction::RemoteScrollingCoordinatorTransaction):
(WebKit::RemoteScrollingCoordinatorTransaction::frameIdentifier const):
(WebKit::RemoteScrollingCoordinatorTransaction::setFrameIdentifier):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::commitScrollingTreeState):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
(WebKit::RemoteScrollingTree::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::waitForEventDefaultHandlingCompletion):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::buildTransaction):

Canonical link: https://commits.webkit.org/274918@main
